### DISCUSSION TITLE
Installation Scripts deprecated or not?

### DISCUSSION THREAD

**Discussion by thisis-akash** [nodesource/distributions#1639] [Q&A] [answered]
In the Readme it says that the installation scripts are no longer supported but they are still downloadable for all versions
https://deb.nodesource.com/setup_14.x
https://deb.nodesource.com/setup_16.x
https://deb.nodesource.com/setup_18.x
https://deb.nodesource.com/setup_20.x

What am I missing here? Are they still valid? If not, why are they available to download? Can installing from them cause an issue?

**Comment by collimarco** [comment]
Also the migration process is not very clear... Waiting for clarifications. Why are they being deprecated? Why not just update them?

**Reply by thisis-akash** [reply]
Well according to the doc, it says the scripts are no longer supported _"as the installation process is straightforward for any RPM and DEB distro"_

**Reply by collimarco** [reply]
So, why not keep them with just the 5-10 lines of code required? They would still be straightforward and useful and backward compatible...

**Reply by JesusPaz** [reply]
Hello @collimarco , what is not clear about the guide? the scripts are not updated because we changed the repository and created a new one with more features that were always asked for. If something is not clear to you, let us know, in the same way you can participate and help us improve the guides.

**Reply by riosje** [reply]
Hi @rdev32 
>`the new instructions say that you have to build your own deb package (?)`.

absolutely not, this process would be much more complex than execute 3 command lines.

What is **create a Deb repo**? is just add a file with some info into `/etc/apt/sources.list.d` thats what the commands do, create the file `nodesource.list` with one line of config.
just execute the commands in the instructions order, why are you stuck in the step 2?

**Reply by rdev32** [reply]
@riosje yes I'm sorry didn't explained myself
I noticed the instructions doesn't work on WSL specifically on the part `NODE_MAJOR=20`
i don't know bash but that seems like a way to define variables. So I tested it and well it doesn't work on wsl
<img width="218" alt="image" src="https://github.com/nodesource/distributions/assets/78289548/f448baa7-3262-44d0-a796-97da0161f082">
this explains because 20 is not appended in `https://deb.nodesource.com/node_$NODE_MAJOR.x`
<img width="634" alt="image" src="https://github.com/nodesource/distributions/assets/78289548/a0430487-ae8d-473d-99f9-a471caacb90e">
guess i have to delete my distro and re do everything again because the uninstall instructions wont remove the files `/etc/apt/keyrings/nodesource.gpg` `/etc/apt/sources.list.d/nodesource.list`
im loosing it man why won't ubuntu ship the latest lts version of node?

**Reply by riosje** [reply]
@rdev32 it seems to be working the variables on your shell.
try this
```
NODE_MAJOR=20
echo $NODE_MAJOR
```

Any way if it doesn't work just replace the `$NODE_MAJOR` placeholder in the URL like this:
```bash
echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_20.x nodistro main" | sudo tee /etc/apt/sources.list.d/nodesource.list
```

We use that variable to avoid write the same thing multiple times for each version, which I think is silly but a lot of people struggle with this, We will reconsider create the instruction for each one.

**Reply by rdev32** [reply]
@riosje thanks for your patience I remade the installation of wsl just for starting fresh and did everything logged as root. Everything worked flawlessly I really don't know why the first time was giving me errors.
Maybe I made typos of some kind
Sorry for bothering my understandings were incorrect

**Comment by riosje** [comment, marked as answer]
Hi everyone,

We understand that changes can sometimes be frustrating for some users, which is why we've created this space to discuss these changes and provide clarity.

### About Script Availability

In our README, we've mentioned:

> Installation Scripts: The installation scripts setup_XX.x are no longer supported and are not needed anymore, as the installation process is straightforward for any RPM and DEB distro.

It's essential to clarify that "no longer supported" doesn't mean the scripts are "unavailable" or "not accessible." Instead, these scripts are now deprecated from our end, and we won't provide support for them. Additionally, users who continue to install Node.js using these scripts won't receive the latest Node.js versions.

As mentioned by @collimarco, one approach would be to refactor the scripts to use the new repository. However, let me explain why we believe documenting the installation process is a better choice for everyone:

1. **Communication:** It's crucial to inform our users about the updates to our repositories and processes.

2. **Issue Resolution:** From time to time, users encounter issues related to package managers like APT or YUM or other Linux installation problems. Some of these users open issues claiming that the scripts are broken, even when the issues aren't script-related.

3. **Clarity:** We prefer that users understand the installation process and how our repository works. The script-based installation method isn't clear for everyone, while the actual installation process consists of just three steps for Debian and one for Fedora.

We believe that documenting the installation process provides transparency and empowers our users to make informed decisions. If you have any questions or suggestions, please feel free to share them.

Thank you for your understanding and collaboration.

**Reply by thisis-akash** [reply]
Okay. One suggestion though, there should be more elaboration on this in the doc. If the script is not dependable/reliable anymore, it should better outlined along with the likely outcome if the script is used. Best case scenario still being the script not be accessible anymore IMHO. I had to spent a lot of time to figure out that the source of the problem I was facing was due to the script now being deprecated. If you had not replied as fast as you did and mentioned that the users using the script wont receive the latest node version, I would have reached the same conclusion after much more effort from my end.

**Reply by riosje** [reply]
but what kind of problem you had? because the scripts remains working as it was...

Remove the scripts is not an option, it will break a lot of pipelines and other stuff. 
When you execute the script it should be displaying this warning.
![image](https://github.com/nodesource/distributions/assets/55195249/114856fb-3396-416f-9ad1-669cb9de0d80)

**Reply by thisis-akash** [reply]
The big warning is surely a nice to have but it can be missed and does not suffice IMHO. 

To give you more context into my problem as you asked, I had this big install script I was working on. So you run this script to install the software system on the linux machine. It produces a lot of logs (thus warning is easy to miss, even the bright red color did not come up but after seeing your screenshot, I recall that I did see something like this), and one of the tasks it has is to purge and remove node and then reinstall node using the installation script. The script was written long back when this must have been valid. My task was to install node version based on the underlying ubuntu version, thus make the process conditional. So I updated the code and now it was installing node using the script URL which had the new node version in its URL but it still kept installing an older node version as I deduced from the logs and checking the node version after the script completed execution. 

When I added a apt cache clean command and ran this 'sudo rm -r /etc/apt/sources.list.d/nodesource.list' after I read a few things around it, it then somehow installed an even older node version, v12. There were other dependencies that the script managed which depended upon a specific node version too.  For all of this I had to run that huge script multiple times. I checked out this repo and read that line that said installation scripts are no longer supported but I could see that the scripts were accessible and so I did not have sufficient information about whether the script is valid or not. For that I would have to interpret and understand what the script is trying to do which I believe is an unneeded overhead at my end. I then decided to start a discussion and proceed with skimming through my source code , taking a lot of time, to see if there is something else causing the script to install an older version of node and I could not find it in the source code as I suspected and thus just when I finally decided that it had to be related to the script being unsupported, your comment came in and confirmed the issue.

And I am a little confused too, your first response mentions that the user using the script will not get the latest version of node and then you mentioned that the scripts remain working as they were? I mean I fetched and ran the setup script for node v16 and v20. It did get them but it failed to install them. As described above, it first installed v14 and then v12.

I hope you can see the reasons behind my suggestions by reading my experience and decision plan as described above.

**Reply by lukeca362** [reply]
So how do we refactor this, so that we no longer use the setup scripts?
    curl -sLO https://deb.nodesource.com/setup_18.x && \
    bash ./setup_18.x && \
    apt-get -y install --no-install-recommends nodejs

**Reply by q0rban** [reply]
@lukeca362 if you follow the new [installation instructions](https://github.com/nodesource/distributions/blob/69a45587cd87bd8c700e40bb8a8160e0c28d71d8/README.md#installation-instructions) you would do the following (I removed the `sudo`s since it looks like your script is running as root already):

```
apt-get update
apt-get install -y ca-certificates curl gnupg
mkdir -p /etc/apt/keyrings
curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg
NODE_MAJOR=18
echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_$NODE_MAJOR.x nodistro main" > /etc/apt/sources.list.d/nodesource.list
apt-get update
apt-get -y install --no-install-recommends nodejs
```

**Reply by chakra-gold** [reply]
@q0rban thanks.

**Comment by creativegrow-dev** [comment]
Pessoal, estou pasando por problemas em meus scripts devido a essa atualização.


ÉSTA ERA A MANEIRA QUE EU ESTAVA INSTALANDO O NODE

`curl -fsSL https://deb.nodesource.com/setup_16.x | sudo -E bash -

sudo apt-get install -y nodejs

npm install -g npm@latest

npm install -g pm2@latest`

Não estou conseguindo instalar e nem executar o node e npm

alguem tem um exemplo aí pra me salvar??

**Reply by riosje** [reply]
Hi @ronandesign please describe your issue in English so anyone can provide support.

**Reply by Deepak-RefactorQ** [reply]
Hi @riosje , What is the simplest way to install nodejs after this update?

**Reply by riosje** [reply]
Hi @Deepak-RefactorQ the easiest way is following the Readme instructions.

**Comment by chester23491** [comment]
Hi maintainers,

We used this script in our CICD pipelines to add node during the build process. We always knew this may be not the best solution, since it use a lot of resources. However, you created a real pain point with the one minute delay showing the deprecated message, slowing down the process.

Since I never had any issues with the scripts I'm quite surprised about your motivations removing the script in the future. As my example shows, our pipelines relay on the script and I'm not alone with this use case.

We decided to build our own build image including the node installation like this:

```Dockerfile
FROM mcr.microsoft.com/dotnet/sdk:6.0

# Install node
ENV NODE_MAJOR 18
RUN apt-get update && apt-get install -y \
    ca-certificates \
    curl \
    gnupg \
    && mkdir -p /etc/apt/keyrings \
    && curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg \
    && echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_${NODE_MAJOR}.x nodistro main" | tee /etc/apt/sources.list.d/nodesource.list \
    && apt update && apt install -y nodejs

```

I hope this solution will help others!

Regards,
Charles from Earlybyte GmbH

**Reply by riosje** [reply]
Hi @chester23491.
I'm glad to see that you've migrated out from the scripts.
This still prove the whole point, those scripts are no needed, just creates overhead to the maintainers.

**Comment by negalaz** [comment]
Can I install node js 14 this way?

**Reply by riosje** [reply]
Unfortunately we do not support Node14 any more.
You can try to use our old scripts for this, but we will not guarantee success or bring support.
https://github.com/nodesource/distributions/blob/master/OLD_README.md

**Reply by lachesis** [reply]
Ah rats, this is a big problem for us. We use node 14 in a lot of places. NodeSource has always been incredibly helpful for us because of its very simple 2-line installation process. How can we continue to install node 14 without receiving a 60 second delay in CI?

Edit: here's an answer, though I do not like it:  https://github.com/nodesource/distributions/discussions/1678

Still happy to hear better answers!

**Reply by negalaz** [reply]
Hi, Idk if using NVM would be a problem for you.

**Comment by morpig** [comment]
If an 8 line command install as enduser is considered "straightforward" compared to a 2 line install, I don't know what to say....

**Reply by riosje** [reply]
You can make the installation process one liner command if you wish...
What we mean with **"**straightforward**"** is that there is no a "hidden" logic behind a script with 350 lines of code that nobody cares about, script which failed a lot of times due to specific user configs.

In the other hand our new instructions is Just a step by step to config a DEB repo, is not like we're asking the users to build Nodejs from scratch, is just a DEB repo, in the same way as you have to config any other DEB repo.

Also you can just put the step by step into your own bash script and then will have the one liner Nodejs installer.

**Comment by thisis-akash** [comment]
![image](https://github.com/nodesource/distributions/assets/30195464/6db27186-661a-44f3-9670-b021e9106355)

If the gpg key is there why is this happening again and again.

Following are some of the errors I was facing previously. I resolved some of them but still stuck on the one mentioned above.

curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | sudo gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg
curl: /usr/lib/libcurl.so.4: no version information available (required by curl)
curl: symbol lookup error: curl: undefined symbol: curl_mime_headers, version CURL_OPENSSL_4
File '/etc/apt/keyrings/nodesource.gpg' exists. Overwrite? (y/N) Y
gpg: no valid OpenPGP data found.

There is also this error I am facing which says its not able to verify some public key:
Hit:6 https://deb.nodesource.com/node_20.x nodistro InRelease
Err:6 https://deb.nodesource.com/node_20.x nodistro InRelease
The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 2F59B5F99B1BE0B4
Ign:7 https://dl.google.com/linux/chrome/deb stable InRelease
Ign:8 https://ppa.launchpadcontent.net/git-core/ppa/ubuntu jammy InRelease
Hit:9 https://ppa.launchpadcontent.net/ricotz/docky/ubuntu jammy InRelease
Hit:8 https://ppa.launchpadcontent.net/git-core/ppa/ubuntu jammy InRelease
Hit:7 https://dl.google.com/linux/chrome/deb stable InRelease
Reading package lists... Done
W: An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: https://deb.nodesource.com/node_20.x nodistro InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 2F59B5F99B1BE0B4
W: Failed to fetch https://deb.nodesource.com/node_20.x/dists/nodistro/InRelease The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 2F59B5F99B1BE0B4
W: Some index files failed to download. They have been ignored, or old ones used instead.

**Comment by seo-rii** [comment]
I understand that the previous script was deprecated, but I think a minute's delay should be deleted after posting a warning. The reasons are as follows:

1. for novice developers
This change is a huge barrier to preventing the influx of node.js novice developers. As it was, you can start node.js right away by just doing a proper Google search and copying and pasting two lines. But what? Enter code and set environment variables? What else is the GPG key error? Look at the Python. apt-get install python3. See how to install "modern" again node.js. I really don't think this is a "simple" and efficient way to bear that level of discomfort. Right now, haven't you answered several issues related to setting environmental variables? They could ask questions because they knew what GitHub was, but even more beginners would try to study other languages such as Python rather than ask about it.
2. Legacy Code Pipeline
Projects that rely on legacy code will get one minute of delay for each build until they eliminate the dependence on that legacy code.

The warning log alone is sufficient to make the library fall into the deprecated state. I think deprecation, where security vulnerabilities are not the reason, should not interfere with the build pipeline.

After I finished writing and read it again, I think it was written a little aggressively, so I apologize. I understand the purpose of the change in installation, and I don't think it's wrong. I'm just saying that I'm wondering if there's a need to break the old way that's working well.

**Reply by morpig** [reply]
its our fault for using this script^^

i have to manually update script in 20+ servers because of the delay problem. and installing new one is tiring, because you need to copy more lines + gpg keys etcetc.

i suggest that you try the following install script which is only one line!! the project itself is open source (https://github.com/Schniz/fnm)

for ubuntu:
```
curl -fsSL https://fnm.vercel.app/install | bash && source ~/.bashrc && fnm install --latest
```

for macos:
```
curl -fsSL https://fnm.vercel.app/install | bash && source ~/.zshrc && fnm install --latest
```

**Comment by GauriSpears** [comment]
What is new way to install the latest node.js without determining NODE_MAJOR? I used setup_current.x before and don't want to care which version is latest.

**Reply by GauriSpears** [reply]
After all I found that the easiest way to install the latest version of Node.JS is https://github.com/tj/n (RTFM - yes, you can use it without npm/before installing npm).
Nvm is good too, but its installation is rather complicated, especially if you'd like to install it for all users.
Hope this will help someone else.

**Comment by hamidh2** [comment]
I had the below script on my docker file for installing nodejs
```
# Install NodeJs
RUN apt-get update && \
apt-get install -y wget && \
apt-get install -y gnupg2 && \
wget -qO- https://deb.nodesource.com/setup_18.x | bash - && \
apt-get install -y build-essential nodejs
# End Install
```
 I noticed that it's not working, after lot sof search and reading comments I noticed I can fix it by below code 
 
 ```
 # Install NodeJs
RUN apt-get update && \
    apt-get install -y ca-certificates curl gnupg && \
    curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/trusted.gpg.d/nodesource.gpg && \
    NODE_MAJOR=20 && \
    echo "deb [signed-by=/etc/apt/trusted.gpg.d/nodesource.gpg] https://deb.nodesource.com/node_$NODE_MAJOR.x nodistro main" | tee /etc/apt/sources.list.d/nodesource.list && \
    apt-get update && \
    apt-get install -y nodejs
# End Install
 ```

**Reply by Haprog** [reply]
Thanks. I ended up doing something very similar but I wanted to keep the commands more consistent with what is documented here https://github.com/nodesource/distributions#installation-instructions
and I wanted to have the Node version be declared clearly on a separate line and taking advantage of layer caching a bit more when changing Node version so I ended up with this:

```Dockerfile
# Install NodeJS
RUN apt-get update && \
    apt-get install -y ca-certificates curl gnupg && \
    mkdir -p /etc/apt/keyrings && \
    curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg
ARG NODE_MAJOR=18
RUN echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_$NODE_MAJOR.x nodistro main" | tee /etc/apt/sources.list.d/nodesource.list && \
    apt-get update && \
    apt-get install nodejs -y
```

I needed this to get Node in a .NET project where the base image is `mcr.microsoft.com/dotnet/sdk`

**Comment by chobo2** [comment]
Is this way no longer working? We have it on our road map to move to the "new" way of doing this, but no timeline was given when "installation scripts" would stop working.

Today I am seeing https://deb.nodesource.com/node_18.x bullseye/main amd64 nodejs amd64 18.17.1-deb-1nodesource1 is coming back with a 404.

**Reply by brentquackenbush** [reply]
Yup getting the same 404. [hamidh2](https://github.com/hamidh2) provided the fix. I don't see where to declare the NODE_MAJOR to be lts though. Seems like I have to now manually choose the NodeJs version.

**Reply by chobo2** [reply]
Hmm, would have been good if they had updated their website to let people know.  Let me know if you figure out where NODE_MAJOR is being declared.

**Comment by Capez** [comment]
`node: --openssl-legacy-provider is not allowed in NODE_OPTIONS` is the version 18 no longer supporting NODE_OPTIONS? this wasnt an issue with the 18 setup script
